File: examples/javascript/matrixtraversalExamples.js
// 1. Spiral Matrix
function spiralOrder(matrix) {
    if (!matrix || matrix.length === 0) return [];
    
    const result = [];
    let top = 0, bottom = matrix.length - 1;
    let left = 0, right = matrix[0].length - 1;
    
    while (top <= bottom && left <= right) {
        // Traverse right
        for (let col = left; col <= right; col++) {
            result.push(matrix[top][col]);
        }
        top++;
        
        // Traverse down
        for (let row = top; row <= bottom; row++) {
            result.push(matrix[row][right]);
        }
        right--;
        
        // Traverse left
        if (top <= bottom) {
            for (let col = right; col >= left; col--) {
                result.push(matrix[bottom][col]);
            }
            bottom--;
        }
        
        // Traverse up
        if (left <= right) {
            for (let row = bottom; row >= top; row--) {
                result.push(matrix[row][left]);
            }
            left++;
        }
    }
    
    return result;
}

// 2. Rotate Image
function rotate(matrix) {
    const n = matrix.length;
    
    // Transpose
    for (let i = 0; i < n; i++) {
        for (let j = i; j < n; j++) {
            [matrix[i][j], matrix[j][i]] = [matrix[j][i], matrix[i][j]];
        }
    }
    
    // Reverse each row
    for (let i = 0; i < n; i++) {
        matrix[i].reverse();
    }
}

// 3. Set Matrix Zeroes
function setZeroes(matrix) {
    const m = matrix.length, n = matrix[0].length;
    let firstRowZero = false, firstColZero = false;
    
    // Check if first row has zero
    for (let j = 0; j < n; j++) {
        if (matrix[0][j] === 0) {
            firstRowZero = true;
            break;
        }
    }
    
    // Check if first column has zero
    for (let i = 0; i < m; i++) {
        if (matrix[i][0] === 0) {
            firstColZero = true;
            break;
        }
    }
    
    // Mark zeros in first row and column
    for (let i = 1; i < m; i++) {
        for (let j = 1; j < n; j++) {
            if (matrix[i][j] === 0) {
                matrix[i][0] = 0;
                matrix[0][j] = 0;
            }
        }
    }
    
    // Set zeros based on marks
    for (let i = 1; i < m; i++) {
        for (let j = 1; j < n; j++) {
            if (matrix[i][0] === 0 || matrix[0][j] === 0) {
                matrix[i][j] = 0;
            }
        }
    }
    
    // Set first row to zero if needed
    if (firstRowZero) {
        for (let j = 0; j < n; j++) {
            matrix[0][j] = 0;
        }
    }
    
    // Set first column to zero if needed
    if (firstColZero) {
        for (let i = 0; i < m; i++) {
            matrix[i][0] = 0;
        }
    }
}

// 4. Word Search
function exist(board, word) {
    const m = board.length, n = board[0].length;
    
    function backtrack(i, j, k) {
        if (k === word.length) return true;
        if (i < 0 || i >= m || j < 0 || j >= n || board[i][j] !== word[k]) {
            return false;
        }
        
        const temp = board[i][j];
        board[i][j] = '#';
        
        const found = backtrack(i + 1, j, k + 1) ||
                     backtrack(i - 1, j, k + 1) ||
                     backtrack(i, j + 1, k + 1) ||
                     backtrack(i, j - 1, k + 1);
        
        board[i][j] = temp;
        return found;
    }
    
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (backtrack(i, j, 0)) return true;
        }
    }
    return false;
}

// 5. Number of Islands
function numIslands(grid) {
    if (!grid || grid.length === 0) return 0;
    
    const m = grid.length, n = grid[0].length;
    let count = 0;
    
    function dfs(i, j) {
        if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] === '0') {
            return;
        }
        
        grid[i][j] = '0';
        dfs(i + 1, j);
        dfs(i - 1, j);
        dfs(i, j + 1);
        dfs(i, j - 1);
    }
    
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (grid[i][j] === '1') {
                count++;
                dfs(i, j);
            }
        }
    }
    
    return count;
}

// 6. Surrounded Regions
function solve(board) {
    if (!board || board.length === 0) return;
    
    const m = board.length, n = board[0].length;
    
    function dfs(i, j) {
        if (i < 0 || i >= m || j < 0 || j >= n || board[i][j] !== 'O') {
            return;
        }
        
        board[i][j] = 'T';
        dfs(i + 1, j);
        dfs(i - 1, j);
        dfs(i, j + 1);
        dfs(i, j - 1);
    }
    
    // Mark boundary-connected 'O's
    for (let i = 0; i < m; i++) {
        dfs(i, 0);
        dfs(i, n - 1);
    }
    for (let j = 0; j < n; j++) {
        dfs(0, j);
        dfs(m - 1, j);
    }
    
    // Convert remaining 'O's to 'X' and 'T's back to 'O'
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (board[i][j] === 'O') {
                board[i][j] = 'X';
            } else if (board[i][j] === 'T') {
                board[i][j] = 'O';
            }
        }
    }
}

// 7. Pacific Atlantic Water Flow
function pacificAtlantic(heights) {
    if (!heights || heights.length === 0) return [];
    
    const m = heights.length, n = heights[0].length;
    const pacific = Array.from({ length: m }, () => new Array(n).fill(false));
    const atlantic = Array.from({ length: m }, () => new Array(n).fill(false));
    
    function dfs(i, j, ocean, prevHeight) {
        if (i < 0 || i >= m || j < 0 || j >= n || 
            ocean[i][j] || heights[i][j] < prevHeight) {
            return;
        }
        
        ocean[i][j] = true;
        dfs(i + 1, j, ocean, heights[i][j]);
        dfs(i - 1, j, ocean, heights[i][j]);
        dfs(i, j + 1, ocean, heights[i][j]);
        dfs(i, j - 1, ocean, heights[i][j]);
    }
    
    // Start from Pacific borders
    for (let i = 0; i < m; i++) {
        dfs(i, 0, pacific, 0);
        dfs(i, n - 1, atlantic, 0);
    }
    for (let j = 0; j < n; j++) {
        dfs(0, j, pacific, 0);
        dfs(m - 1, j, atlantic, 0);
    }
    
    const result = [];
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (pacific[i][j] && atlantic[i][j]) {
                result.push([i, j]);
            }
        }
    }
    
    return result;
}

// 8. Game of Life
function gameOfLife(board) {
    const m = board.length, n = board[0].length;
    const directions = [[-1,-1],[-1,0],[-1,1],[0,-1],[0,1],[1,-1],[1,0],[1,1]];
    
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            let liveNeighbors = 0;
            
            for (const [di, dj] of directions) {
                const ni = i + di, nj = j + dj;
                if (ni >= 0 && ni < m && nj >= 0 && nj < n && 
                    (board[ni][nj] === 1 || board[ni][nj] === 2)) {
                    liveNeighbors++;
                }
            }
            
            if (board[i][j] === 1 && (liveNeighbors < 2 || liveNeighbors > 3)) {
                board[i][j] = 2; // Live -> Dead
            }
            if (board[i][j] === 0 && liveNeighbors === 3) {
                board[i][j] = 3; // Dead -> Live
            }
        }
    }
    
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (board[i][j] === 2) board[i][j] = 0;
            if (board[i][j] === 3) board[i][j] = 1;
        }
    }
}

// 9. Valid Sudoku
function isValidSudoku(board) {
    const rows = Array.from({ length: 9 }, () => new Set());
    const cols = Array.from({ length: 9 }, () => new Set());
    const boxes = Array.from({ length: 9 }, () => new Set());
    
    for (let i = 0; i < 9; i++) {
        for (let j = 0; j < 9; j++) {
            const val = board[i][j];
            if (val === '.') continue;
            
            const boxIndex = Math.floor(i / 3) * 3 + Math.floor(j / 3);
            
            if (rows[i].has(val) || cols[j].has(val) || boxes[boxIndex].has(val)) {
                return false;
            }
            
            rows[i].add(val);
            cols[j].add(val);
            boxes[boxIndex].add(val);
        }
    }
    
    return true;
}

// 10. Shortest Path in Binary Matrix
function shortestPathBinaryMatrix(grid) {
    const n = grid.length;
    if (grid[0][0] === 1 || grid[n-1][n-1] === 1) return -1;
    if (n === 1) return 1;
    
    const queue = [[0, 0, 1]];
    const visited = new Set(['0,0']);
    const directions = [[-1,-1],[-1,0],[-1,1],[0,-1],[0,1],[1,-1],[1,0],[1,1]];
    
    while (queue.length > 0) {
        const [row, col, path] = queue.shift();
        
        for (const [dr, dc] of directions) {
            const newRow = row + dr;
            const newCol = col + dc;
            const key = `${newRow},${newCol}`;
            
            if (newRow >= 0 && newRow < n && newCol >= 0 && newCol < n && 
                grid[newRow][newCol] === 0 && !visited.has(key)) {
                
                if (newRow === n-1 && newCol === n-1) return path + 1;
                
                visited.add(key);
                queue.push([newRow, newCol, path + 1]);
            }
        }
    }
    
    return -1;
}

// Test cases
console.log("=== Matrix Traversal Examples ===");

const matrix = [[1,2,3],[4,5,6],[7,8,9]];
console.log("Spiral Order:", spiralOrder(matrix));

const rotateMatrix = [[1,2,3],[4,5,6],[7,8,9]];
rotate(rotateMatrix);
console.log("Rotated Matrix:", rotateMatrix);

const board = [['A','B','C','E'],['S','F','C','S'],['A','D','E','E']];
console.log("Word Search 'ABCCED':", exist(board, "ABCCED"));

const grid = [['1','1','0','0','0'],['1','1','0','0','0'],['0','0','1','0','0'],['0','0','0','1','1']];
console.log("Number of Islands:", numIslands(grid));

const heights = [[1,2,2,3,5],[3,2,3,4,4],[2,4,5,3,1],[6,7,1,4,5],[5,1,1,2,4]];
console.log("Pacific Atlantic cells:", pacificAtlantic(heights).length);

const binaryGrid = [[0,0,0],[1,1,0],[1,1,0]];
console.log("Shortest Path Binary Matrix:", shortestPathBinaryMatrix(binaryGrid));

// Export functions
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        spiralOrder,
        rotate,
        setZeroes,
        exist,
        numIslands,
        solve,
        pacificAtlantic,
        gameOfLife,
        isValidSudoku,
        shortestPathBinaryMatrix
    };
}

File: examples/python/matrix_traversal_examples.py
# 1. Spiral Order
def spiral_order(matrix):
    if not matrix or not matrix[0]:
        return []
    
    result = []
    top, bottom = 0, len(matrix) - 1
    left, right = 0, len(matrix[0]) - 1
    
    while top <= bottom and left <= right:
        # Traverse right
        for col in range(left, right + 1):
            result.append(matrix[top][col])
        top += 1
        
        # Traverse down
        for row in range(top, bottom + 1):
            result.append(matrix[row][right])
        right -= 1
        
        # Traverse left
        if top <= bottom:
            for col in range(right, left - 1, -1):
                result.append(matrix[bottom][col])
            bottom -= 1
        
        # Traverse up
        if left <= right:
            for row in range(bottom, top - 1, -1):
                result.append(matrix[row][left])
            left += 1
    
    return result

# 2. Rotate Matrix
def rotate_matrix(matrix):
    n = len(matrix)
    
    # Transpose
    for i in range(n):
        for j in range(i, n):
            matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
    
    # Reverse each row
    for i in range(n):
        matrix[i].reverse()

# 3. Set Zeroes
def set_zeroes(matrix):
    m, n = len(matrix), len(matrix[0])
    first_row_zero = any(matrix[0][j] == 0 for j in range(n))
    first_col_zero = any(matrix[i][0] == 0 for i in range(m))
    
    # Use first row and column as markers
    for i in range(1, m):
        for j in range(1, n):
            if matrix[i][j] == 0:
                matrix[i][0] = 0
                matrix[0][j] = 0
    
    # Set zeros based on markers
    for i in range(1, m):
        for j in range(1, n):
            if matrix[i][0] == 0 or matrix[0][j] == 0:
                matrix[i][j] = 0
    
    # Handle first row and column
    if first_row_zero:
        for j in range(n):
            matrix[0][j] = 0
    
    if first_col_zero:
        for i in range(m):
            matrix[i][0] = 0

# 4. Search Matrix
def search_matrix(matrix, target):
    if not matrix or not matrix[0]:
        return False
    
    m, n = len(matrix), len(matrix[0])
    row, col = 0, n - 1
    
    while row < m and col >= 0:
        if matrix[row][col] == target:
            return True
        elif matrix[row][col] > target:
            col -= 1
        else:
            row += 1
    
    return False

# 5. Num Islands
def num_islands(grid):
    if not grid or not grid[0]:
        return 0
    
    m, n = len(grid), len(grid[0])
    count = 0
    
    def dfs(i, j):
        if i < 0 or i >= m or j < 0 or j >= n or grid[i][j] != '1':
            return
        grid[i][j] = '0'
        dfs(i + 1, j)
        dfs(i - 1, j)
        dfs(i, j + 1)
        dfs(i, j - 1)
    
    for i in range(m):
        for j in range(n):
            if grid[i][j] == '1':
                count += 1
                dfs(i, j)
    
    return count

# 6. Max Area Island
def max_area_island(grid):
    if not grid or not grid[0]:
        return 0
    
    m, n = len(grid), len(grid[0])
    max_area = 0
    
    def dfs(i, j):
        if i < 0 or i >= m or j < 0 or j >= n or grid[i][j] != 1:
            return 0
        grid[i][j] = 0
        return 1 + dfs(i + 1, j) + dfs(i - 1, j) + dfs(i, j + 1) + dfs(i, j - 1)
    
    for i in range(m):
        for j in range(n):
            if grid[i][j] == 1:
                max_area = max(max_area, dfs(i, j))
    
    return max_area

# 7. Surrounded Regions
def surrounded_regions(board):
    if not board or not board[0]:
        return
    
    m, n = len(board), len(board[0])
    
    def dfs(i, j):
        if i < 0 or i >= m or j < 0 or j >= n or board[i][j] != 'O':
            return
        board[i][j] = 'T'  # Temporary mark
        dfs(i + 1, j)
        dfs(i - 1, j)
        dfs(i, j + 1)
        dfs(i, j - 1)
    
    # Mark border-connected 'O's
    for i in range(m):
        dfs(i, 0)
        dfs(i, n - 1)
    for j in range(n):
        dfs(0, j)
        dfs(m - 1, j)
    
    # Convert remaining 'O's to 'X' and restore 'T's to 'O'
    for i in range(m):
        for j in range(n):
            if board[i][j] == 'O':
                board[i][j] = 'X'
            elif board[i][j] == 'T':
                board[i][j] = 'O'

# 8. Pacific Atlantic
def pacific_atlantic(heights):
    if not heights or not heights[0]:
        return []
    
    m, n = len(heights), len(heights[0])
    pacific = [[False] * n for _ in range(m)]
    atlantic = [[False] * n for _ in range(m)]
    
    def dfs(i, j, ocean, prev_height):
        if (i < 0 or i >= m or j < 0 or j >= n or 
            ocean[i][j] or heights[i][j] < prev_height):
            return
        
        ocean[i][j] = True
        dfs(i + 1, j, ocean, heights[i][j])
        dfs(i - 1, j, ocean, heights[i][j])
        dfs(i, j + 1, ocean, heights[i][j])
        dfs(i, j - 1, ocean, heights[i][j])
    
    # Start from Pacific borders
    for i in range(m):
        dfs(i, 0, pacific, heights[i][0])
    for j in range(n):
        dfs(0, j, pacific, heights[0][j])
    
    # Start from Atlantic borders
    for i in range(m):
        dfs(i, n - 1, atlantic, heights[i][n - 1])
    for j in range(n):
        dfs(m - 1, j, atlantic, heights[m - 1][j])
    
    result = []
    for i in range(m):
        for j in range(n):
            if pacific[i][j] and atlantic[i][j]:
                result.append([i, j])
    
    return result

# 9. Shortest Path Binary Matrix
def shortest_path_binary_matrix(grid):
    n = len(grid)
    if grid[0][0] == 1 or grid[n-1][n-1] == 1:
        return -1
    
    if n == 1:
        return 1
    
    from collections import deque
    queue = deque([(0, 0, 1)])
    visited = {(0, 0)}
    directions = [(-1,-1), (-1,0), (-1,1), (0,-1), (0,1), (1,-1), (1,0), (1,1)]
    
    while queue:
        row, col, path_length = queue.popleft()
        
        for dr, dc in directions:
            nr, nc = row + dr, col + dc
            
            if (0 <= nr < n and 0 <= nc < n and 
                grid[nr][nc] == 0 and (nr, nc) not in visited):
                
                if nr == n - 1 and nc == n - 1:
                    return path_length + 1
                
                visited.add((nr, nc))
                queue.append((nr, nc, path_length + 1))
    
    return -1

# 10. Walls And Gates
def walls_and_gates(rooms):
    if not rooms or not rooms[0]:
        return
    
    m, n = len(rooms), len(rooms[0])
    from collections import deque
    queue = deque()
    
    # Find all gates
    for i in range(m):
        for j in range(n):
            if rooms[i][j] == 0:
                queue.append((i, j))
    
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    
    while queue:
        row, col = queue.popleft()
        
        for dr, dc in directions:
            nr, nc = row + dr, col + dc
            
            if (0 <= nr < m and 0 <= nc < n and 
                rooms[nr][nc] == 2147483647):  # INF
                rooms[nr][nc] = rooms[row][col] + 1
                queue.append((nr, nc))

# 11. Unique Paths With Obstacles
def unique_paths_with_obstacles(obstacle_grid):
    if not obstacle_grid or obstacle_grid[0][0] == 1:
        return 0
    
    m, n = len(obstacle_grid), len(obstacle_grid[0])
    dp = [[0] * n for _ in range(m)]
    dp[0][0] = 1
    
    for i in range(m):
        for j in range(n):
            if obstacle_grid[i][j] == 1:
                dp[i][j] = 0
            else:
                if i > 0:
                    dp[i][j] += dp[i-1][j]
                if j > 0:
                    dp[i][j] += dp[i][j-1]
    
    return dp[m-1][n-1]

# 12. Min Path Sum
def min_path_sum(grid):
    m, n = len(grid), len(grid[0])
    
    for i in range(1, m):
        grid[i][0] += grid[i-1][0]
    
    for j in range(1, n):
        grid[0][j] += grid[0][j-1]
    
    for i in range(1, m):
        for j in range(1, n):
            grid[i][j] += min(grid[i-1][j], grid[i][j-1])
    
    return grid[m-1][n-1]

# 13. Word Search
def word_search(board, word):
    if not board or not board[0]:
        return False
    
    m, n = len(board), len(board[0])
    
    def dfs(i, j, k):
        if k == len(word):
            return True
        if i < 0 or i >= m or j < 0 or j >= n or board[i][j] != word[k]:
            return False
        
        temp = board[i][j]
        board[i][j] = '#'  # Mark as visited
        
        found = (dfs(i+1, j, k+1) or dfs(i-1, j, k+1) or 
                dfs(i, j+1, k+1) or dfs(i, j-1, k+1))
        
        board[i][j] = temp  # Restore
        return found
    
    for i in range(m):
        for j in range(n):
            if dfs(i, j, 0):
                return True
    
    return False

# 14. Run Examples
def run_examples():
    print("=== Matrix Traversal Examples ===")
    
    # Test 1: Spiral Order
    matrix = [[1,2,3],[4,5,6],[7,8,9]]
    print(f"Spiral Order: {spiral_order(matrix)}")
    
    # Test 2: Rotate Matrix
    matrix = [[1,2,3],[4,5,6],[7,8,9]]
    rotate_matrix(matrix)
    print(f"Rotated Matrix: {matrix}")
    
    # Test 3: Search Matrix
    matrix = [[1,4,7,11],[2,5,8,12],[3,6,9,16],[10,13,14,17]]
    print(f"Search Matrix (5): {search_matrix(matrix, 5)}")
    
    # Test 4: Number of Islands
    grid = [["1","1","1","1","0"],["1","1","0","1","0"],["1","1","0","0","0"],["0","0","0","0","0"]]
    print(f"Number of Islands: {num_islands(grid)}")
    
    # Test 5: Max Area of Island
    grid = [[0,0,1,0,0,0,0,1,0,0,0,0,0],[0,0,0,0,0,0,0,1,1,1,0,0,0],[0,1,1,0,1,0,0,0,0,0,0,0,0]]
    print(f"Max Area Island: {max_area_island(grid)}")
    
    # Test 6: Shortest Path Binary Matrix
    grid = [[0,0,0],[1,1,0],[1,1,0]]
    print(f"Shortest Path Binary Matrix: {shortest_path_binary_matrix(grid)}")
    
    # Test 7: Unique Paths with Obstacles
    obstacle_grid = [[0,0,0],[0,1,0],[0,0,0]]
    print(f"Unique Paths with Obstacles: {unique_paths_with_obstacles(obstacle_grid)}")
    
    # Test 8: Min Path Sum
    grid = [[1,3,1],[1,5,1],[4,2,1]]
    print(f"Min Path Sum: {min_path_sum(grid)}")
    
    # Test 9: Word Search
    board = [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]]
    word = "ABCCED"
    print(f"Word Search '{word}': {word_search(board, word)}")
    
    # Test 10: Pacific Atlantic Water Flow
    heights = [[1,2,2,3,5],[3,2,3,4,4],[2,4,5,3,1],[6,7,1,4,5],[5,1,1,2,4]]
    print(f"Pacific Atlantic: {pacific_atlantic(heights)}")

if __name__ == "__main__":
    run_examples()